File: java/udp/ServidorUDP.java
import java.net.*;
import java.time.LocalDateTime;

public class ServidorUDP {
    public static void main(String[] args) throws Exception {
        int porta = 5030;
        byte[] buffer = new byte[1024];

        try (DatagramSocket socket = new DatagramSocket(porta)) {
            System.out.println("[UDP] Servidor aguardando datagramas na porta " + porta + "...");
            while (true) {
                DatagramPacket pacoteRecebido = new DatagramPacket(buffer, buffer.length);
                socket.receive(pacoteRecebido);

                String mensagem = new String(pacoteRecebido.getData(), 0, pacoteRecebido.getLength());
                System.out.println("[UDP] Recebido de " + pacoteRecebido.getAddress() + ": " + mensagem);

                String resposta;
                if (mensagem.equalsIgnoreCase("hora")) {
                    resposta = "Horário do servidor: " + LocalDateTime.now();
                } else {
                    resposta = "Monitor responde: recebi sua mensagem -> \"" + mensagem + "\"";
                }
                byte[] dadosResposta = resposta.getBytes();
                DatagramPacket pacoteResposta = new DatagramPacket(
                        dadosResposta, dadosResposta.length,
                        pacoteRecebido.getAddress(), pacoteRecebido.getPort());
                socket.send(pacoteResposta);
            }
        }
    }
}

File: python/udp/servidor_udp.py
import socket
from datetime import datetime

HOST = "0.0.0.0"
PORTA = 5030

with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as servidor:
    servidor.bind((HOST, PORTA))
    print(f"[UDP] Servidor aguardando datagramas na porta {PORTA}...")

    while True:
        try:
            dados, endereco = servidor.recvfrom(1024)
        except ConnectionResetError:
            continue

        mensagem = dados.decode("utf-8")
        print(f"[UDP] Recebido de {endereco}: {mensagem}")

        if mensagem.lower() == "hora":
            resposta = f"Horário do servidor: {datetime.now()}"
        else:
            resposta = f'Monitor responde: recebi sua mensagem -> "{mensagem}"'

        servidor.sendto(resposta.encode("utf-8"), endereco)